### PR TITLE
samples: wifi: Wi-Fi samples to use Wi-Fi credentials module

### PR DESCRIPTION
**Pull request by D-Triveni**
test_wifi: pr-sdk-nrf-12907

### CODE REVIEW

File: samples/wifi/sr_coex/README.rst [hunk 1]
```diff
@@ -157,15 +142,18 @@ To enable different test modes, set up the following configuration parameters in
...
   The units are 1.25 milliseconds.
   For example, ``CONFIG_INTERVAL_MIN=80`` corresponds to an interval of 100 ms (80 x 1.25).
-* Wi-Fi connection: Set the following options appropriately as per the credentials of the access point used for this testing:
+* Wi-Fi connection: Set the following Wi-Fi credentials in the :file:`prj.conf` file:
 
-  * :ref:`CONFIG_STA_SAMPLE_SSID <CONFIG_STA_SAMPLE_SSID>`
-  * :ref:`CONFIG_STA_SAMPLE_PASSWORD <CONFIG_STA_SAMPLE_PASSWORD>`
-  * :ref:`CONFIG_STA_KEY_MGMT_* <CONFIG_STA_KEY_MGMT_*>`
-  * :kconfig:option:`CONFIG_NET_CONFIG_PEER_IPV4_ADDR`
+* :kconfig:option:`CONFIG_WIFI_CREDENTIALS_STATIC` - This option enables static Wi-Fi configuration.
```
Comment (krish2718) [inline samples/wifi/sr_coex/README.rst:147]: Why can't we just refer to the `wifi_credentials` doc here? Checkout `options-from-kconfig` Sphinx extension in NCS.

File: samples/wifi/sr_coex/src/main.c [hunk 2]
```diff
@@ -194,21 +194,21 @@ static int __wifi_args_to_params(struct wifi_connect_req_params *params)
 	params->timeout = SYS_FOREVER_MS;
 
 	/* SSID */
-	params->ssid = CONFIG_STA_SAMPLE_SSID;
+	params->ssid = CONFIG_WIFI_CREDENTIALS_STATIC_SSID;
 	params->ssid_length = strlen(params->ssid);
 
-#if defined(CONFIG_STA_KEY_MGMT_WPA2)
+#if defined(CONFIG_WIFI_CREDENTIALS_STATIC_TYPE_PSK)
```
Comment (krish2718) [inline samples/wifi/sr_coex/src/main.c:200]: This can be a helper function in the library

File: samples/wifi/sr_coex/prj.conf [hunk 3]
```diff
@@ -70,13 +70,15 @@ CONFIG_NET_BUF_TX_COUNT=48
...
-# CONFIG_STA_KEY_MGMT_NONE=y
-# CONFIG_STA_KEY_MGMT_WPA2=y
-# CONFIG_STA_KEY_MGMT_WPA2_256=y
-# CONFIG_STA_KEY_MGMT_WPA3=y
-CONFIG_STA_SAMPLE_SSID="Myssid"
-CONFIG_STA_SAMPLE_PASSWORD="Mypassword"
+# CONFIG_WIFI_CREDENTIALS_STATIC_TYPE_OPEN=y
+# CONFIG_WIFI_CREDENTIALS_STATIC_TYPE_PSK=y
+# CONFIG_WIFI_CREDENTIALS_STATIC_TYPE_PSK_SHA256=y
+# CONFIG_WIFI_CREDENTIALS_STATIC_TYPE_SAE=y
```
Comment (krish2718) [inline samples/wifi/sr_coex/prj.conf:79]: Do we need all these mentioned under comments? Just set WPA2/WPA3 as default.

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: samples/wifi/sr_coex/prj.conf (hunk 3 above)
Comment (D-Triveni) [reply samples/wifi/sr_coex/prj.conf:79]: WPA2 is used as default in library Kconfig, so using the same here and removing other configs.

Comment (krish2718) [reply samples/wifi/sr_coex/prj.conf:79]: Then we can remove it altogether.

File: samples/wifi/sr_coex/README.rst (hunk 1 above)
Comment (D-Triveni) [reply samples/wifi/sr_coex/README.rst:147]: ``wifi_credentials_options.txt`` is being included in all other samples to list ``Wi-Fi credentials``. Doing the same here.

Comment (krish2718) [reply samples/wifi/sr_coex/README.rst:147]: Sure, but that itself won't be enough

1. Add a heading for `configuration options`
2. If `Kconfig` then use `options-from-kconfig` for that sample
3. Then add Wi-Fi credentials: `.. include:: /includes/wifi_credentials_options.txt`

File: samples/wifi/sr_coex/README.rst [hunk 4]
```diff
@@ -71,70 +71,8 @@ Configuration options
...
-.. _CONFIG_BT_THROUGHPUT_FILE:
-
-CONFIG_BT_THROUGHPUT_FILE
-   This option selects the type of the throughput test.
-
-.. _CONFIG_BT_THROUGHPUT_DURATION:
-
-CONFIG_BT_THROUGHPUT_DURATION
-   This option sets the Bluetooth throughput test duration in milliseconds.
+.. options-from-kconfig::
```
Comment (krish2718) [inline samples/wifi/sr_coex/README.rst:74]: it's better if this is in a separate commit before we switch to Wi-Fi crendetials.

File: samples/wifi/twt/src/main.c [hunk 5]
```diff
@@ -42,6 +42,9 @@ struct traffic_gen_config tg_config;
 static struct net_mgmt_event_callback wifi_shell_mgmt_cb;
 static struct net_mgmt_event_callback net_shell_mgmt_cb;
 
+int __wifi_creds_to_params(size_t argc, char *argv[],
```
Comment (krish2718) [inline samples/wifi/twt/src/main.c:45]: Why is this not in a header file `include/net/wifi_credentials.h`

File: subsys/net/lib/wifi_credentials/wifi_credentials.c [hunk 6]
```diff
@@ -20,6 +22,82 @@ K_MUTEX_DEFINE(wifi_credentials_mutex);
 static char ssid_cache[CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES][WIFI_SSID_MAX_LEN];
 static size_t ssid_cache_lengths[CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES];
 
+int __wifi_creds_to_params(size_t argc, char *argv[],
```
Comment (krish2718) [inline subsys/net/lib/wifi_credentials/wifi_credentials.c:25]: Why do we need variable args here? Can't we pass a simple `struct`? And also, except for Password/Passphrase others are straightforward, so, helper can be limited to that.

File: subsys/net/lib/wifi_credentials/wifi_credentials.c [hunk 7]
```diff
@@ -20,6 +22,82 @@ K_MUTEX_DEFINE(wifi_credentials_mutex);
...
+		}
+
+		if (params->channel == 0U) {
+			params->channel = WIFI_CHANNEL_ANY;
+		}
+
+		idx++;
+	} else {
+		params->channel = WIFI_CHANNEL_ANY;
+	}
```
Comment (krish2718) [inline subsys/net/lib/wifi_credentials/wifi_credentials.c:56]: This is nothing to do with wifi_credentials

**Review by krish2718 — CHANGES_REQUESTED** [review]

**Comment by D-Triveni** [comment]
Previously, the default key management was set to SAE, and the AP used for CI tests was also configured accordingly. However, we have now changed the key management to PSK. As a result, the CI tests are failing, which was expected due to this change.

**Review by krish2718 — APPROVED** [review]

**Review by bama-nordic — APPROVED** [review]

**Review by richabp — CHANGES_REQUESTED** [review]

**Review by muraliThokala — APPROVED** [review]

**Review by richabp — APPROVED** [review]

**Review by bama-nordic — APPROVED** [review]

File: samples/wifi/sr_coex/README.rst [hunk 8]
```diff
@@ -71,90 +71,28 @@ Configuration options
...
 
-* Antenna configuration: Use the :ref:`CONFIG_COEX_SEP_ANTENNAS <CONFIG_COEX_SEP_ANTENNAS>` Kconfig option to select the antenna configuration.
+* Antenna configuration: Use the :kconfig:option:`CONFIG_COEX_SEP_ANTENNAS` Kconfig option to select the antenna configuration.
   Set it to ``y`` to enable separate antennas and ``n`` to enable shared antenna.
 * Test modes: Use the following Kconfig options to select the required test case:
 
-  * :ref:`CONFIG_TEST_TYPE_WLAN_ONLY <CONFIG_TEST_TYPE_WLAN_ONLY>` for Wi-Fi only test
-  * :ref:`CONFIG_TEST_TYPE_BLE_ONLY <CONFIG_TEST_TYPE_BLE_ONLY>` for Bluetooth LE only test
-  * :ref:`CONFIG_TEST_TYPE_WLAN_BLE <CONFIG_TEST_TYPE_WLAN_BLE>` for concurrent Wi-Fi and Bluetooth LE test.
+  * :kconfig:option:`CONFIG_TEST_TYPE_WLAN_ONLY` for Wi-Fi only test
```
Comment (krish2718) [inline samples/wifi/sr_coex/README.rst:86]: can you please move these changes to a separate commit?

File: include/net/wifi_credentials.h [hunk 9]
```diff
@@ -192,6 +194,15 @@ typedef void (*wifi_credentials_ssid_cb)(void *cb_arg, const char *ssid, size_t
  */
 void wifi_credentials_for_each_ssid(wifi_credentials_ssid_cb cb, void *cb_arg);
 
+
+/**
+ * @brief Converts Wi-Fi credentials configured via Kconfigs into Wi-Fi management params.
+ *
+ * @param params	Pointer to struct in which Wi-Fi connect request params are stored
+ */
+int wifi_credentials_load(struct wifi_connect_req_params *params);
```
Comment (krish2718) [inline include/net/wifi_credentials.h:203]: This should be a separate commit.

File: subsys/net/lib/wifi_credentials/wifi_credentials.c [hunk 10]
```diff
@@ -20,6 +21,32 @@ K_MUTEX_DEFINE(wifi_credentials_mutex);
 static char ssid_cache[CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES][WIFI_SSID_MAX_LEN];
 static size_t ssid_cache_lengths[CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES];
 
+int wifi_credentials_load(struct wifi_connect_req_params *params)
```
Comment (krish2718) [inline subsys/net/lib/wifi_credentials/wifi_credentials.c:24]: `wifi_credentials_load_static`

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: samples/wifi/sr_coex/README.rst (hunk 8 above)
Comment (D-Triveni) [reply samples/wifi/sr_coex/README.rst:86]: Build errors were seen related to the `` :ref:`` label after using ``:show_type:`` extension. So, addressed the issues by making all the necessary changes in a single commit.

File: subsys/net/lib/wifi_credentials/wifi_credentials.c [hunk 11]
```diff
@@ -20,6 +21,32 @@ K_MUTEX_DEFINE(wifi_credentials_mutex);
 static char ssid_cache[CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES][WIFI_SSID_MAX_LEN];
 static size_t ssid_cache_lengths[CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES];
 
+int wifi_credentials_load_static(struct wifi_connect_req_params *params)
+{
+#ifdef CONFIG_WIFI_CREDENTIALS_STATIC
```
Comment (gregersrygg) [inline subsys/net/lib/wifi_credentials/wifi_credentials.c:26]: Please change the `#ifdef`s to use the [IS_ENABLED](https://docs.zephyrproject.org/apidoc/latest/group__sys-util.html#ga111fe4e9d63758262fc6810a782cb32a) macro to get cleaner code and better error messages.

**Review by gregersrygg — APPROVED** [review]

File: subsys/net/lib/wifi_credentials/wifi_credentials.c (hunk 11 above)
Comment (D-Triveni) [reply subsys/net/lib/wifi_credentials/wifi_credentials.c]: @gregersrygg 
Modified code to use ``if IS_ENABLED(CONFIG_WIFI_CREDENTIALS_STATIC)`` instead of 
``if (IS_ENABLED(CONFIG_WIFI_CREDENTIALS_STATIC))``
This change is to fix the compilation errors for applications that do not enable 
 ``CONFIG_WIFI_CREDENTIALS_STATIC`` option.

Comment (gregersrygg) [reply subsys/net/lib/wifi_credentials/wifi_credentials.c]: Did you get compilation errors? Could it be another reason? The `IS_ENABLED` macro is supposed to handle the case when the value is not defined at all: https://docs.zephyrproject.org/apidoc/latest/group__sys-util.html#ga111fe4e9d63758262fc6810a782cb32a

Using an `#if` macro is actually discouraged in the Zephyr docs because it doesn't give any benefits vs plain `#if defined()`

Comment (D-Triveni) [reply subsys/net/lib/wifi_credentials/wifi_credentials.c]: Yes, compilation errors were seen in the not defined case , as he `` CONFIG_WIFI_CREDENTIALS_STATIC_SSID`` and ``CONFIG_WIFI_CREDENTIALS_STATIC_PSK`` are available only if ``CONFIG_WIFI_CREDENTIALS_STATIC`` option is defined.
If ``CONFIG_WIFI_CREDENTIALS_STATIC`` option is not enabled, C compiler is still trying to compile the 
if (IS_ENBALED(config)) block and then the errors were seen.

File: include/net/wifi_credentials.h [hunk 12]
```diff
@@ -192,6 +194,14 @@ typedef void (*wifi_credentials_ssid_cb)(void *cb_arg, const char *ssid, size_t
  */
 void wifi_credentials_for_each_ssid(wifi_credentials_ssid_cb cb, void *cb_arg);
 
+/**
+ * @brief Converts Wi-Fi credentials configured via Kconfigs into Wi-Fi management params.
+ *
+ * @param params       Pointer to struct in which Wi-Fi connect request params are stored
+ */
+int wifi_credentials_load_static(struct wifi_connect_req_params *params);
```
Comment (krish2718) [inline include/net/wifi_credentials.h:202]: offline teams comment: Instead of this API, use `NET_REQUEST_WIFI_CONNECT_STORED` net_mgmt call

**Review by bjda — APPROVED** [review]

**Review by gregersrygg — COMMENTED** [review]
Changes are fine by me. It seems to remove some unrelated kconfig options, but I assume that is intended. Will leave it to other reviewers to approve.

**Review by krish2718 — APPROVED** [review]

**Review by bama-nordic — APPROVED** [review]

**Review by VivekUppunda — APPROVED** [review]

**Review by krga2022 — APPROVED** [review]

**Review by richabp — APPROVED** [review]
